rascal-lsp/src/main/rascal/library/util/LanguageServer.rsc

```diff
@@ -279,6 +279,109 @@ data LanguageService
+        For example if the user starts completion like this: `pr|i` (where `|` represents the cursor) and you generate the completion edit to
+        replace the whole string `pri` with `print`, *filterText* should be set to `pr` or the completion will not show up in the list.
+        If you leave this blank, the correct filter text will be generated for you. If not, we assume you know what you are doing and use the
+        *filterText* you provide as-is.
+    * *deprecated*: Set this to `true` to mark this completion suggestion as deprecated.
+    * *commitCharacters*: When one of these characters is types, the completion suggestion is accepted and the the commit character is inserted after the inserted text.
+    * *additionalChanges*: Any additional changes anywhere else in the document. This can for instance be used to add imports.
+}
+data CompletionSuggestion = completion(
+    CompletionKind kind,
+    CompletionEdit edit,
+
+    str label,
+    str labelDetail = "",
+    str labelDescription = "",
+
+    str details = "",
+    CompletionDocumentation documentation = none(),
+
+    str sortText = "",
+    str filterText = "",
+
+    bool deprecated = false,
+    bool preselect = false,
+
+    list[str] commitCharacters = [],
+
+    list[TextEdit] additionalChanges = []
+);
+
+@description{
+    Definition of a completion edit:
+    * *startColumn (required): The column where the completion edit operation will take place. Must be at or before the cursor position.
+    * *insertEndColumn* (required): End column when the user choose completion by insertion  (for instance by pressing "Enter" in VSCode).
+    * *replaceEndColumn* (required): End column when the user choose completion by replacement (for instance by pressing "Shift-Enter" in VSCode).
+        Note: *insertEndColumn* must not be larger than *replaceEndColumn* and both must be at or to the right of the cursor position.
+    * *newText* *required*: The text that will be used to perform the completion. Depending on what kind of completion (insertion
+        or replacement) selected by the user, the original text from *startColumn* to either *insertEndColumn* or *replaceEndColumn* will
+        be replaced by *newText*.
+    * *snippet*: Can be set to true to indicate that the replacement text should be interpreted as a snippet.
+        Snippets can contain tabstops, placeholders, choices, and variables. For more information about snippets see the
+        [relevant section of the lsp specification on github](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#snippet_syntax).
+        Note that it is the responsibility of the language developer that snippets as presented to the user are syntactically correct to
+        provide a decent user experience.
+}
+data CompletionEdit = completionEdit(
+    int startColumn,
+    int insertEndColumn,
+    int replaceEndColumn,
```

Comment:
Can we not do multi-line completions?

Reply:
Answered by the [spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem):
> *Note:* The range of the edit must be a single line range and it must contain the position at which completion has been requested.